File: C++/HeapTest.cpp
#include "HeapTest.h"
#include "HeapApp.h"
#include <vector>
#include <iostream>
#include <cassert>

using namespace std;

void test_merge_arrays() {
    vector<vector<int>> v;
    v.push_back({1,5,6,9,11});
    v.push_back({23,45,56,79,192});
    v.push_back({2,4,8,10,12});
    v.push_back({15,20,35, 40});
    vector<int> merged = merge_arrays(v);
    for (int val : merged) {
        cout << val << " ";
    }
    cout << endl;
}

void test_sort_k_increasing_decreasing_array() {
    vector<int> a = {57,131,493,294,221,339,418,452,442,190};
    vector<int> b = sort_k_increasing_descreasing_array(a);
    for (int val: b) {
        cout << val << " ";
    }
    cout << endl;
}

void test_heap_stack_queue() {
    HeapStack<int> stack;
    stack.push(1);
    stack.push(2);
    stack.push(3);
    assert(stack.top() == 3);
    stack.pop();
    assert(stack.top() == 2);
    stack.pop();
    assert(stack.top() == 1);
    stack.pop();
    assert(stack.empty());
    
    HeapQueue<int> queue;
    queue.push(1);
    queue.push(2);
    queue.push(3);
    assert(queue.front() == 1);
    queue.pop();
    assert(queue.front() == 2);
    queue.pop();
    assert(queue.front() == 3);
    queue.pop();
    assert(queue.empty());
}

void test_k_closest_stars() {
    string stringvalues ="0,0.0,0.0,0.0\n1,0.0,0.0,1.0\n2,1.0,0.0,0.0\n3,0.0,1.0,0.0\n4,0.0,1.0,0.0\n5,3.0,1.0,0.0\n6,3.0,1.0,2.0\n7,8.0,1.0,9.0\n8,2.0,1.0,0.0\n";  
    istringstream ss(stringvalues);
    vector<Star> closest_stars = find_closest_k_stars(ss, 4);
    for (Star star : closest_stars) {
        cout << star.ID << " ";
    }
    cout << endl;
}

void test_kth_largest_stream() {
    cout << "kth largerst stream" << endl;
    string stringvalues = "10 5 3 2 1 6 9 11 23 17 99";
    istringstream ss(stringvalues);
    find_kth_largest_stream<int>(ss, 3);
}

void test_approximate_sort() {
    cout << "Approximate sort: " << endl;
    string stringvalues = "10 5 3 7 23 33 99 80 100 60";
    istringstream ss(stringvalues);
    approximate_sort<int>(ss, 4);    
}

void test_k_closest_median() {
    cout << "k closest median" << endl;
    vector<int> v = {7, 14, 10, 12, 2, 11, 29, 3, 4};
    vector<int> u = find_k_closest_to_median(v, 5);
    for (int i : u) {
        cout << i << endl;
    }
}

void test_online_median() {
    cout << "Online median" << endl;
    string values = "1 2 3 5 10 9 8 7";
    istringstream ss(values);
    online_median<int>(ss);
}

void test_generate_first_k() {
    cout << "Generate first k" << endl;
    vector<Num> v = generate_first_k(10);
    for (Num n: v) {
        cout << n.val_ << endl;
    }
}

void test_compare_kth_largest() {
    vector<int> v = {16, 14, 10, 8, 7, 9, 3, 2, 4, 1};
    assert(compare_k_th_largest_heap(v, 3, 9) == 1);
    assert(compare_k_th_largest_heap(v, 3, 10) == 0);
    assert(compare_k_th_largest_heap(v, 3, 11) == -1);
}

void test_heap() {
    test_merge_arrays();
    test_sort_k_increasing_decreasing_array();
    test_heap_stack_queue();
    test_k_closest_stars();
    test_kth_largest_stream();
    test_approximate_sort();
    test_k_closest_median();
    test_online_median();
    test_generate_first_k();
    test_compare_kth_largest();
}

File: src/HeapTest.java
import org.junit.*;
import static org.junit.Assert.*;

import java.io.*;
import java.util.*;

public class HeapTest {
	@Test
	public void test_merge_arrays() {
		Integer[][] array = {{1,5,7}, {3,6,9}, {4, 8, 10}};
		List<Integer> merged = HeapUtil.merge_arrays(array);
		for (Integer val: merged) {
			System.out.print(val + " ");
		}
		System.out.println();
	}
	
	@Test
	public void test_sort_k_up_down_array() {
		/* Test case 1 */
		Integer[] a = {57, 131, 493, 294, 221, 339, 418, 452, 442, 190};
		Integer[] b = HeapUtil.sort_k_up_down_array(a);
		assertEquals(b[0].intValue(), 57);
		assertEquals(b[1].intValue(), 131);
		assertEquals(b[2].intValue(), 190);
		assertEquals(b[3].intValue(), 221);
		assertEquals(b[4].intValue(), 294);
		assertEquals(b[5].intValue(), 339);
		assertEquals(b[6].intValue(), 418);
		assertEquals(b[7].intValue(), 442);
		assertEquals(b[8].intValue(), 452);
		assertEquals(b[9].intValue(), 493);
		
		/*Test case 2 */
		Integer[] c = {5, 6, 4, 3, 8, 9, 1, 2, 7};
		Integer[] d = HeapUtil.sort_k_up_down_array(c);
		assertEquals(d[0].intValue(), 1);
		assertEquals(d[1].intValue(), 2);
		assertEquals(d[2].intValue(), 3);
		assertEquals(d[3].intValue(), 4);
		assertEquals(d[4].intValue(), 5);
		assertEquals(d[5].intValue(), 6);
		assertEquals(d[6].intValue(), 7);
		assertEquals(d[7].intValue(), 8);
		assertEquals(d[8].intValue(), 9);
	}
	
	@Test
	public void test_Heap_Stack_Queue() throws Exception{
		/* Test for stack */ 
		HeapStack<Integer> stack = new HeapStack();
		stack.push(1);
		stack.push(4);
		stack.push(10);
		assertEquals(stack.pop().intValue(), 10);
		assertEquals(stack.pop().intValue(), 4);
		assertEquals(stack.pop().intValue(), 1);
		
		/* Test for queue */
		HeapQueue<Integer> queue = new HeapQueue();
		queue.enqueue(1);
		queue.enqueue(4);
		queue.enqueue(10);
		assertEquals(queue.dequeue().intValue(), 1);
		assertEquals(queue.dequeue().intValue(), 4);
		assertEquals(queue.dequeue().intValue(), 10);		
	}
	
	@Test
	public void test_k_closest_stars() {
		List<Star> starList = new ArrayList<Star>();
		starList.add(new Star(0, 0.0, 0.0, 0.0));
		starList.add(new Star(1, 0.0, 0.0, 1.0));
		starList.add(new Star(2, 1.0, 0.0, 0.0));
		starList.add(new Star(3, 0.0, 1.0, 0.0));
		starList.add(new Star(4, 0.0, 1.0, 0.0));
		starList.add(new Star(5, 3.0, 1.0, 0.0));
		starList.add(new Star(6, 3.0, 1.0, 2.0));
		starList.add(new Star(7, 8.0, 1.0, 9.0));
		starList.add(new Star(8, 2.0, 1.0, 0.0));
		starList.add(new Star(9, -3.0, 1.0, -6.0));
		List<Star> closestStars = Star.k_closest_stars(starList, 4);
		for (Star star: closestStars) {
			System.out.print(star.getId() + " ");
		}
		System.out.println();
	}
	
	@Test
	public void test_kth_largest_item() throws FileNotFoundException, IOException, ClassNotFoundException {		
		String inputFileName = "/home/thanh/workspace/EPI/input/kth_largest_item.dat";
		String outputFileName = "/home/thanh/workspace/EPI/output/kth_largest_item.out";
		Integer[] list = {13, 23, 1, 4, 5, 6, 20, 100, 2, 3};
		/* Write to the input file */
		ObjectOutputStream stream = new ObjectOutputStream(new FileOutputStream(inputFileName));
		for (Integer i: list) {
			stream.writeObject(i);
		}
		stream.flush();
		ObjectInputStream input = new ObjectInputStream(new FileInputStream(inputFileName));
		FileOutputStream output = new FileOutputStream(outputFileName);
		HeapUtil.print_kth_largest_item(input, output, 3);
	}
	
	@Test
	public void test_appoximate_sort() {
		List<Integer> list = Arrays.asList(10, 2, 3, 20, 50, 100, 34, 99, 80);
		List<Integer> sortedList = HeapUtil.approximate_sort(list, 3);
		Iterator<Integer> it = sortedList.iterator();
		assertEquals(it.next().intValue(), 2);
		assertEquals(it.next().intValue(), 3);
		assertEquals(it.next().intValue(), 10);
		assertEquals(it.next().intValue(), 20);
		assertEquals(it.next().intValue(), 34);
		assertEquals(it.next().intValue(), 50);
		assertEquals(it.next().intValue(), 80);
		assertEquals(it.next().intValue(), 99);
		assertEquals(it.next().intValue(), 100);
	}
	
	@Test
	public void test_Median_Heap() {
		int[] input = {2,3,4,1,5,7,10,9};
		MedianHeap<Integer> heap = new MedianHeap<Integer>();
		for (int i: input) {
			heap.add(i);
			System.out.print(heap.getMedian() + " ");
		}
		System.out.println();
	}
	
	@Test
	public void test_kth_largest_heap() {
		Integer[] max_heap = {16, 14, 10, 8, 7, 9, 3, 2, 4, 1};
		assertEquals(HeapUtil.compare_kth_largest_heap(max_heap, 3, 10), 0);
		assertEquals(HeapUtil.compare_kth_largest_heap(max_heap, 3, 9), 1);
		assertEquals(HeapUtil.compare_kth_largest_heap(max_heap, 3, 11), -1);
	}
}
